### ISSUE TITLE
[Bug] The inference results obtained using the SDK are different from the preview results.

### ISSUE THREAD

**Issue by mogazheng** [open-mmlab/mmdeploy#2104]
### Checklist

- [X] I have searched related issues but cannot get the expected help.
- [X] 2. I have read the [FAQ documentation](https://github.com/open-mmlab/mmdeploy/tree/main/docs/en/faq.md) but cannot get the expected help.
- [X] 3. The bug has not been fixed in the latest version.

### Describe the bug

codebase: MMOCR
model: DBNetpp

When exporting the TensorRT model by executing `tools/deploy.py` , the preview image of the inference results is shown in Figure 1. However, when performing inference using the SDK after exporting the model information using dump-info, the output in Figure 2 differs. The bounding box in Figure 2 is larger.
Why do the results differ between the two?

here is the preview images，dump info and model
[Google Drive Link](https://drive.google.com/drive/folders/1KbwWxOJtEp--T5ZlpMaVV_DVlT0_rA6g?usp=sharing)

### Reproduction

```
import cv2
from mmdeploy_runtime import TextDetector

img = cv2.imread("0001_00-02-01.jpg")
detector = TextDetector(model_path="trt_with_dump_info", device_name="cuda", device_id=0)
bboxes = detector(img)
if len(bboxes) > 0:
    pts = (bboxes[:, 0:8].reshape(len(bboxes), -1, 2).astype(int))
    cv2.polylines(img, pts, True, (0, 255, 0), 2)
    cv2.imshow('Figure_2', img)
    cv2.waitKey(0)
```

### Environment

```Shell
05/20 17:38:57 - mmengine - INFO - **********Environmental information**********
05/20 17:39:31 - mmengine - INFO - sys.platform: win32
05/20 17:39:31 - mmengine - INFO - Python: 3.10.9 | packaged by Anaconda, Inc. | (main, Mar  8 2023, 10:42:25) [MSC 
v.1916 64 bit (AMD64)]
05/20 17:39:31 - mmengine - INFO - CUDA available: True
05/20 17:39:31 - mmengine - INFO - numpy_random_seed: 2147483648
05/20 17:39:31 - mmengine - INFO - GPU 0: NVIDIA GeForce RTX 3060 Laptop GPU
05/20 17:39:31 - mmengine - INFO - CUDA_HOME: C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v11.6
05/20 17:39:31 - mmengine - INFO - NVCC: Cuda compilation tools, release 11.6, V11.6.124
05/20 17:39:31 - mmengine - INFO - MSVC: 用于 x64 的 Microsoft (R) C/C++ 优化编译器 19.29.30148 版
05/20 17:39:31 - mmengine - INFO - GCC: n/a
05/20 17:39:31 - mmengine - INFO - PyTorch: 1.12.1+cu116
05/20 17:39:31 - mmengine - INFO - PyTorch compiling details: PyTorch built with:
  - C++ Version: 199711
  - MSVC 192829337
  - Intel(R) Math Kernel Library Version 2020.0.2 Product Build 20200624 for Intel(R) 64 architecture applications  
  - Intel(R) MKL-DNN v2.6.0 (Git Hash 52b5f107dd9cf10910aaa19cb47f3abf9b349815)
  - OpenMP 2019
  - LAPACK is enabled (usually provided by MKL)
  - CPU capability usage: AVX2
  - CUDA Runtime 11.6
  - NVCC architecture flags: -gencode;arch=compute_37,code=sm_37;-gencode;arch=compute_50,code=sm_50;-gencode;arch=compute_60,code=sm_60;-gencode;arch=compute_61,code=sm_61;-gencode;arch=compute_70,code=sm_70;-gencode;arch=compute_75,code=sm_75;-gencode;arch=compute_80,code=sm_80;-gencode;arch=compute_86,code=sm_86;-gencode;arch=compute_37,code=compute_37
  - CuDNN 8.3.2  (built against CUDA 11.5)
  - Magma 2.5.4
  - Build settings: BLAS_INFO=mkl, BUILD_TYPE=Release, CUDA_VERSION=11.6, CUDNN_VERSION=8.3.2, CXX_COMPILER=C:/actions-runner/_work/pytorch/pytorch/builder/windows/tmp_bin/sccache-cl.exe, CXX_FLAGS=/DWIN32 /D_WINDOWS /GR /EHsc /w /bigobj -DUSE_PTHREADPOOL -openmp:experimental -IC:/actions-runner/_work/pytorch/pytorch/builder/windows/mkl/include -DNDEBUG -DUSE_KINETO -DLIBKINETO_NOCUPTI -DUSE_FBGEMM -DUSE_XNNPACK -DSYMBOLICATE_MOBILE_DEBUG_HANDLE -DEDGE_PROFILER_USE_KINETO, LAPACK_INFO=mkl, PERF_WITH_AVX=1, PERF_WITH_AVX2=1, PERF_WITH_AVX512=1, TORCH_VERSION=1.12.1, USE_CUDA=ON, USE_CUDNN=ON, USE_EXCEPTION_PTR=1, USE_GFLAGS=OFF, USE_GLOG=OFF, USE_MKL=ON, USE_MKLDNN=OFF, USE_MPI=OFF, USE_NCCL=OFF, USE_NNPACK=OFF, USE_OPENMP=ON, USE_ROCM=OFF,

05/20 17:39:31 - mmengine - INFO - TorchVision: 0.13.1+cu116
05/20 17:39:31 - mmengine - INFO - OpenCV: 4.7.0
05/20 17:39:31 - mmengine - INFO - MMEngine: 0.7.0
05/20 17:39:31 - mmengine - INFO - MMCV: 2.0.0rc4
05/20 17:39:31 - mmengine - INFO - MMCV Compiler: MSVC 192829924
05/20 17:39:31 - mmengine - INFO - MMCV CUDA Compiler: 11.6
05/20 17:39:31 - mmengine - INFO - MMDeploy: 1.0.0rc3+ec4abe5
05/20 17:39:31 - mmengine - INFO -

05/20 17:39:31 - mmengine - INFO - **********Backend information**********
05/20 17:39:31 - mmengine - INFO - tensorrt:    8.5.3.1
05/20 17:39:31 - mmengine - INFO - tensorrt custom ops: Available
05/20 17:39:31 - mmengine - INFO - ONNXRuntime: 1.14.1
05/20 17:39:31 - mmengine - INFO - ONNXRuntime-gpu:     1.14.1
05/20 17:39:31 - mmengine - INFO - ONNXRuntime custom ops:      Available
05/20 17:39:31 - mmengine - INFO - pplnn:       None
05/20 17:39:31 - mmengine - INFO - ncnn:        None
05/20 17:39:31 - mmengine - INFO - snpe:        None
05/20 17:39:31 - mmengine - INFO - openvino:    None
05/20 17:39:31 - mmengine - INFO - torchscript: 1.12.1+cu116
05/20 17:39:31 - mmengine - INFO - torchscript custom ops:      NotAvailable
05/20 17:39:31 - mmengine - INFO - rknn-toolkit:        None
05/20 17:39:31 - mmengine - INFO - rknn-toolkit2:       None
05/20 17:39:31 - mmengine - INFO - ascend:      None
05/20 17:39:31 - mmengine - INFO - coreml:      None
05/20 17:39:31 - mmengine - INFO - tvm: None
05/20 17:39:31 - mmengine - INFO - vacc:        None
05/20 17:39:31 - mmengine - INFO -

05/20 17:39:31 - mmengine - INFO - **********Codebase information**********
05/20 17:39:31 - mmengine - INFO - mmdet:       3.0.0rc6
05/20 17:39:31 - mmengine - INFO - mmseg:       None
05/20 17:39:31 - mmengine - INFO - mmcls:       None
05/20 17:39:31 - mmengine - INFO - mmocr:       1.0.0rc6
05/20 17:39:31 - mmengine - INFO - mmedit:      None
05/20 17:39:31 - mmengine - INFO - mmdet3d:     None
05/20 17:39:31 - mmengine - INFO - mmpose:      None
05/20 17:39:31 - mmengine - INFO - mmrotate:    None
05/20 17:39:31 - mmengine - INFO - mmaction:    None
```

**Comment by RunningLeon** [comment]
@AllentDan Could you take a look at this issue?

**Comment by AllentDan** [comment]
Hi, @mogazheng. I retested the model, especially on a dataset ICDAR2015. The result of SDK is equal to or even slightly better than PyTorch results. The implementation of SDK is different from Python since Python uses OpenCV to do pre-process and post-process. While in SDK, if the device is CUDA, the logic could be different as we implemented many methods in our own way. In this aspect, a single image difference is acceptable.

**Comment by mogazheng** [comment]
@AllentDan Thank you very much for your response. I apologize for not providing clear description earlier. Both images were predicted using the same TensorRT model. The first image was visualized using the script `tools/deploy.py`, while the second image was generated using the Python SDK. This issue is not specific to a single sample but occurs in every image within my dataset. Based on my testing, the problem arises with curved text lines, where the predicted bounding box expands more than the ground truth. Since the text lines in the ICDAR dataset are straight, this issue may not exist there. I suggest trying the CTW1500 dataset to see if you can reproduce the problem.

**Comment by AllentDan** [comment]
I tried CTW1500 and your model. The result of PyThon is `icdar/precision: 0.4614  icdar/recall: 0.1466  icdar/hmean: 0.2225` while the result of SDK is `icdar/precision: 0.4948  icdar/recall: 0.1589  icdar/hmean: 0.2405`. The modified model config is:
```python
model = dict(
    type='DBNet',
    backbone=dict(
        type='mmdet.ResNet',
        depth=50,
        num_stages=4,
        out_indices=(0, 1, 2, 3),
        frozen_stages=-1,
        norm_cfg=dict(type='BN', requires_grad=True),
        norm_eval=False,
        style='pytorch',
        dcn=dict(type='DCNv2', deform_groups=1, fallback_on_stride=False),
        init_cfg=dict(type='Pretrained', checkpoint='torchvision://resnet50'),
        stage_with_dcn=(False, True, True, True)),
    neck=dict(
        type='FPNC',
        in_channels=[256, 512, 1024, 2048],
        lateral_channels=256,
        asf_cfg=dict(attention_type='ScaleChannelSpatial')),
    det_head=dict(
        type='DBHead',
        in_channels=256,
        module_loss=dict(type='DBModuleLoss'),
        postprocessor=dict(
            type='DBPostprocessor', text_repr_type='quad',
            epsilon_ratio=0.002)),
    data_preprocessor=dict(
        type='TextDetDataPreprocessor',
        mean=[123.675, 116.28, 103.53],
        std=[58.395, 57.12, 57.375],
        bgr_to_rgb=True,
        pad_size_divisor=32))
train_pipeline = [
    dict(type='LoadImageFromFile', color_type='color_ignore_orientation'),
    dict(
        type='LoadOCRAnnotations',
        with_bbox=True,
        with_polygon=True,
        with_label=True),
    dict(
        type='TorchVisionWrapper',
        op='ColorJitter',
        brightness=0.12549019607843137,
        saturation=0.5),
    dict(
        type='ImgAugWrapper',
        args=[['Fliplr', 0.5], {
            'cls': 'Affine',
            'rotate': [-10, 10]
        }, ['Resize', [0.5, 3.0]]]),
    dict(type='RandomCrop', min_side_ratio=0.1),
    dict(type='Resize', scale=(640, 640), keep_ratio=True),
    dict(type='Pad', size=(640, 640)),
    dict(
        type='PackTextDetInputs',
        meta_keys=('img_path', 'ori_shape', 'img_shape'))
]
test_pipeline = [
    dict(type='LoadImageFromFile', color_type='color_ignore_orientation'),
    dict(type='Resize', scale=(2048, 1024), keep_ratio=True),
    dict(
        type='LoadOCRAnnotations',
        with_polygon=True,
        with_bbox=True,
        with_label=True),
    dict(
        type='PackTextDetInputs',
        meta_keys=('img_path', 'ori_shape', 'img_shape', 'scale_factor',
                   'instances'))
]
default_scope = 'mmocr'
env_cfg = dict(
    cudnn_benchmark=False,
    mp_cfg=dict(mp_start_method='fork', opencv_num_threads=0),
    dist_cfg=dict(backend='nccl'))
randomness = dict(seed=None)
default_hooks = dict(
    timer=dict(type='IterTimerHook'),
    logger=dict(type='LoggerHook', interval=5),
    param_scheduler=dict(type='ParamSchedulerHook'),
    checkpoint=dict(type='CheckpointHook', interval=20),
    sampler_seed=dict(type='DistSamplerSeedHook'),
    sync_buffer=dict(type='SyncBuffersHook'),
    visualization=dict(
        type='VisualizationHook',
        interval=1,
        enable=False,
        show=False,
        draw_gt=False,
        draw_pred=False))
log_level = 'INFO'
log_processor = dict(type='LogProcessor', window_size=10, by_epoch=True)
load_from = 'https://download.openmmlab.com/mmocr/textdet/dbnetpp/tmp_1.0_pretrain/dbnetpp_r50dcnv2_fpnc_100k_iter_synthtext-20220502-352fec8a.pth'
resume = False
val_evaluator = dict(type='HmeanIOUMetric')
test_evaluator = dict(type='HmeanIOUMetric')
vis_backends = [dict(type='LocalVisBackend')]
visualizer = dict(
    type='TextDetLocalVisualizer',
    name='visualizer',
    vis_backends=[dict(type='LocalVisBackend')])
ctw1500_textdet_data_root = 'data/ctw1500'
ctw1500_textdet_train = dict(
    type='OCRDataset',
    data_root='data/ctw1500',
    ann_file='textdet_train.json',
    filter_cfg=dict(filter_empty_gt=True, min_size=32),
    pipeline=[
        dict(type='LoadImageFromFile', color_type='color_ignore_orientation'),
        dict(
            type='LoadOCRAnnotations',
            with_bbox=True,
            with_polygon=True,
            with_label=True),
        dict(
            type='TorchVisionWrapper',
            op='ColorJitter',
            brightness=0.12549019607843137,
            saturation=0.5),
        dict(
            type='ImgAugWrapper',
            args=[['Fliplr', 0.5], {
                'cls': 'Affine',
                'rotate': [-10, 10]
            }, ['Resize', [0.5, 3.0]]]),
        dict(type='RandomCrop', min_side_ratio=0.1),
        dict(type='Resize', scale=(640, 640), keep_ratio=True),
        dict(type='Pad', size=(640, 640)),
        dict(
            type='PackTextDetInputs',
            meta_keys=('img_path', 'ori_shape', 'img_shape'))
    ])
ctw1500_textdet_test = dict(
    type='OCRDataset',
    data_root='data/ctw1500',
    ann_file='textdet_test.json',
    test_mode=True,
    pipeline=[
        dict(type='LoadImageFromFile', color_type='color_ignore_orientation'),
        dict(type='Resize', scale=(2048, 1024), keep_ratio=True),
        dict(
            type='LoadOCRAnnotations',
            with_polygon=True,
            with_bbox=True,
            with_label=True),
        dict(
            type='PackTextDetInputs',
            meta_keys=('img_path', 'ori_shape', 'img_shape', 'scale_factor',
                       'instances'))
    ])
optim_wrapper = dict(
    type='OptimWrapper',
    optimizer=dict(type='SGD', lr=0.007, momentum=0.9, weight_decay=0.0001))
train_cfg = dict(type='EpochBasedTrainLoop', max_epochs=1200, val_interval=20)
val_cfg = dict(type='ValLoop')
test_cfg = dict(type='TestLoop')
param_scheduler = [dict(type='PolyLR', power=0.9, eta_min=1e-07, end=1200)]
train_dataloader = dict(
    batch_size=16,
    num_workers=4,
    persistent_workers=True,
    sampler=dict(type='DefaultSampler', shuffle=True),
    dataset=dict(
        type='OCRDataset',
        data_root='data/ctw1500',
        ann_file='textdet_train.json',
        filter_cfg=dict(filter_empty_gt=True, min_size=32),
        pipeline=[
            dict(
                type='LoadImageFromFile',
                color_type='color_ignore_orientation'),
            dict(
                type='LoadOCRAnnotations',
                with_bbox=True,
                with_polygon=True,
                with_label=True),
            dict(
                type='TorchVisionWrapper',
                op='ColorJitter',
                brightness=0.12549019607843137,
                saturation=0.5),
            dict(
                type='ImgAugWrapper',
                args=[['Fliplr', 0.5], {
                    'cls': 'Affine',
                    'rotate': [-10, 10]
                }, ['Resize', [0.5, 3.0]]]),
            dict(type='RandomCrop', min_side_ratio=0.1),
            dict(type='Resize', scale=(640, 640), keep_ratio=True),
            dict(type='Pad', size=(640, 640)),
            dict(
                type='PackTextDetInputs',
                meta_keys=('img_path', 'ori_shape', 'img_shape'))
        ]))
val_dataloader = dict(
    batch_size=1,
    num_workers=4,
    persistent_workers=True,
    sampler=dict(type='DefaultSampler', shuffle=False),
    dataset=dict(
        type='OCRDataset',
        data_root='data/ctw1500',
        ann_file='textdet_test.json',
        test_mode=True,
        pipeline=[
            dict(
                type='LoadImageFromFile',
                color_type='color_ignore_orientation'),
            dict(type='Resize', scale=(2048, 1024), keep_ratio=True),
            dict(
                type='LoadOCRAnnotations',
                with_polygon=True,
                with_bbox=True,
                with_label=True),
            dict(
                type='PackTextDetInputs',
                meta_keys=('img_path', 'ori_shape', 'img_shape',
                           'scale_factor', 'instances'))
        ]))
test_dataloader = dict(
    batch_size=1,
    num_workers=4,
    persistent_workers=True,
    sampler=dict(type='DefaultSampler', shuffle=False),
    dataset=dict(
        type='OCRDataset',
        data_root='data/ctw1500',
        ann_file='textdet_test.json',
        test_mode=True,
        pipeline=[
            dict(
                type='LoadImageFromFile',
                color_type='color_ignore_orientation'),
            dict(type='Resize', scale=(2048, 1024), keep_ratio=True),
            dict(
                type='LoadOCRAnnotations',
                with_polygon=True,
                with_bbox=True,
                with_label=True),
            dict(
                type='PackTextDetInputs',
                meta_keys=('img_path', 'ori_shape', 'img_shape',
                           'scale_factor', 'instances'))
        ]))
auto_scale_lr = dict(base_batch_size=16)

```

**Comment by mogazheng** [comment]
@AllentDan Sorry, I still didn't describe clearly.
This issue is unrelated to either the PyTorch model or the model's accuracy. My model can achieve `icdar/precision: 1.0000 icdar/recall: 1.0000 icdar/hmean: 1.0000` on my validation dataset.

I put it another way.
You can try to execute the following two pieces of code to reproduce this problem.
The first piece of code is taken from the inference visualization section in `tools/deploy.py(line 312)`.

```
from functools import partial
import torch.multiprocessing as mp
from torch.multiprocessing import Process
from mmdeploy.apis import (visualize_model)
from mmdeploy.utils import (get_backend, get_root_logger, load_config, target_wrapper)

def create_process(name, target, args, kwargs, ret_value=None):
    logger = get_root_logger()
    logger.info(f'{name} start.')
    log_level = logger.level

    wrap_func = partial(target_wrapper, target, log_level, ret_value)

    process = Process(target=wrap_func, args=args, kwargs=kwargs)
    process.start()
    process.join()

    if ret_value is not None:
        if ret_value.value != 0:
            logger.error(f'{name} failed.')
            exit(1)
        else:
            logger.info(f'{name} success.')

ret_value = mp.Value('d', 0, lock=False)
deploy_cfg, model_cfg = load_config(
    "text-detection_tensorrt_dynamic-320x320-2240x2240.py", 
    "dbnetpp_resnet50_dcnv2_fpnc_1200e_dh.py")
backend = get_backend(deploy_cfg)
extra = dict(
    backend=backend,
    output_file='Figure_1.jpg',
    show_result=True)
create_process(
    'visualize trt model',
    target=visualize_model,
    args=("dbnetpp_resnet50_dcnv2_fpnc_1200e_dh.py", 
        "text-detection_tensorrt_dynamic-320x320-2240x2240.py", 
        "trt_with_dump_info/end2end.engine", 
        "0001_00-02-01.jpg",
        "cuda"),
    kwargs=extra,
    ret_value=ret_value)
```

and the second piece of code is from the Python SDK demo `mmdeploy/demo/python/ocr.py`. 

```
import cv2
from mmdeploy_runtime import TextDetector

img = cv2.imread("0001_00-02-01.jpg")
detector = TextDetector(model_path="trt_with_dump_info", device_name="cuda", device_id=0)
bboxes = detector(img)
if len(bboxes) > 0:
    pts = (bboxes[:, 0:8].reshape(len(bboxes), -1, 2).astype(int))
    cv2.polylines(img, pts, True, (0, 255, 0), 2)
    cv2.imshow('Figure_2', img)
    cv2.waitKey(0)
```

**Both pieces of code use the converted TRT model.**

The first piece of code outputs the correct result, but the second piece of code outputs the wrong result.

**Comment by AllentDan** [comment]
I totally understand what you meant for. Like I said, the implementation of SDK CUDA part which `TextDetector` applied is different from Python which `create_process` applied. What I tried to express is that the evaluation metric shows that SDK is better even though the predicted boxes are slightly larger than Python results.

Actually, I was debugging in the past few days and found that even though I aligned the pre-process of SDK to the same as Python, the output from TensorRT can just be different, and the difference is about 1.x percentile.

**Comment by mogazheng** [comment]
@AllentDan Thanks for your response.

I think this is not an acceptable problem. 
Because the downstream task of text detection is text recognition, this will have a big impact on text recognition. 
Especially it performs well on straight text lines but encounters problems with every curved text.
If a solution cannot be found, I might have to implement pre-processing and post-processing by myself instead of using SDK pipeline.

**Comment by AllentDan** [comment]
It turns out to be the bug of DBNet pos-process in SDK. I will fix it ASAP.

**Comment by AllentDan** [comment]
Hi, @mogazheng. A quick fix for your case is to modify a line of code [here](https://github.com/open-mmlab/mmdeploy/blob/25cea0ac0f055d7081c696ad1ac129dab1478f4d/csrc/mmdeploy/codebase/mmocr/dbnet.cpp#LL35C10-L35C10). Replace it with `std::string platform = "cpu";`

The bug was raised by [convexHull](https://github.com/open-mmlab/mmdeploy/blob/25cea0ac0f055d7081c696ad1ac129dab1478f4d/csrc/mmdeploy/codebase/mmocr/cuda/dbnet.cpp#L61) can not handle non-convex shape of texts. Temporarily, I can not find out a proper solution to fix it. Hope this workaround can help you out.

**Comment by AllentDan** [comment]
> @AllentDan This workaround is helpful, thank you so much!
> 
> By the way, I noticed that the CPU portion of the code is implemented using OpenCV's `approxPolyDP`. Would it be possible to achieve similar effectiveness by replacing `convexHull` in the CUDA portion with `approxPolyDP`?
> 
> Additionally, the CPU portion does not support the `epsilon_ratio` parameter in MMOCR's DBHead, so modify the code [here](https://github.com/open-mmlab/mmdeploy/blob/25cea0ac0f055d7081c696ad1ac129dab1478f4d/csrc/mmdeploy/codebase/mmocr/cpu/dbnet.cpp#L39) manually is required to obtain identical results as Python.

Could not use `approxPolyDP` since the contour output from CUDA kernel is different from OpenCV. It is disordered and redundant compared to OpenCV result.

As for `epsilon_ratio `, yes it should be decided by config rather than fixed to 0.01. Not a big deal, we may update it in the future versions.
